quodlibet/quodlibet/formats/_audio.py

```diff
@@ -532,26 +533,35 @@ def list(self, key):
             v = self.get(key)
             return [] if v is None else v.split("\n")
 
-    def list_separate(self, key, connector=" - "):
-        """Similar to list, but will return a list of all combinations
-        for tied tags instead of one comma separated string.
-
-        In case of tied tags the result will be unicode, otherwise
-        it returns the same as list()
+    def list_sort(self, key):
+        """Like list but return display,sort pairs when appropriate
+        and work on all tags
         """
+        display = decode_value(key, self.__call__(key))
+        display = display.split("\n") if display else []
+        sort = []
+        if key in TAG_TO_SORT:
+            sort = decode_value(TAG_TO_SORT[key],
+                                self.__call__(TAG_TO_SORT[key]))
+            # it would be better to use something that doesn't fall back
+            # to the key itself, but what?
+            sort = sort.split("\n") if sort else []
+        if len(display) > 0 and len(sort) > 0 and display != sort:
+            return [((d if d != "" else s, s) if s != "" else d)
+                    for d, s in izip_longest(display, sort, fillvalue="")]
+        else:
```

Comment:
I was trying to be somewhat efficient, i.e., if there are no useful sort
values, then don't go through the bother of creating both the display and sort
versions.  The consuming code already handled the mixed situation, because
that is what comes out of the pattern code, so there was no extra downstream
work (except a little bit at the very end) to do things this way.

I also wanted to distinguish between the situation where there is a sort value
and where there isn't so that a subsequent sort value (from a different song,
like where "The Beatles" is sometimes missing the "Beatles, The" sort value)
could be used in the paned browser.

As far as using extra sort values, I was thinking that extra information
should not be dropped.  If they are truely unneeded then they can be removed
from the data.

If you don't think that either of these are useful just let me know and I'll
make the changes.  (Both should be easy.)

peter

PS:  I noticed that using izip doesn't work in Python 3.   Should I include
the compatability magic to make later migration to Python 3 easier?

On 01/22/2016 10:00 AM, Christoph Reiter wrote:
> In quodlibet/quodlibet/formats/_audio.py
> <https://github.com/quodlibet/quodlibet/pull/1796#discussion_r50568841>:
> 
>>          """
>> +        display = decode_value(key, self.__call__(key))
>> +        display = display.split("\n") if display else []
>> +        sort = []
>> +        if key in TAG_TO_SORT:
>> +            sort = decode_value(TAG_TO_SORT[key],
>> +                                self.__call__(TAG_TO_SORT[key]))
>> +            # it would be better to use something that doesn't fall back
>> +            # to the key itself, but what?
>> +            sort = sort.split("\n") if sort else []
>> +        if len(display) > 0 and len(sort) > 0 and display != sort:
>> +            return [((d if d != "" else s, s) if s != "" else d)
>> +                    for d, s in izip_longest(display, sort, fillvalue="")]
>> +        else:
> 
> I think it would be nicer to just always return a list of tuples even if there
> are no sort values. Also throw out sort values if there are too many.
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/quodlibet/quodlibet/pull/1796/files#r50568841>.
>